# Discussion 05

## Agenda: Midterm Review ✍️

* **Understand:** Take a look at Previous midterm patterns
* **Recap:** Revise/Recollect some lecture content related to the midterm
* **Practice:** Solve past year midterm problems for practice

In [17]:
import pandas as pd
import seaborn as sns
import matplotlib
import matplotlib.pyplot as plt
import numpy as np

In [18]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib
import matplotlib.pyplot as plt
import os

# Q1

### The table below shows data collected for 330 penguins in Antarctica. Each row represents one penguin. 

### Three species of penguins appear in the data set: `Adelie`, `Gentoo`, and `Chinstrap`. This data set will be referred to throughout the exam as the DataFrame `df` .

In [19]:
df = pd.read_csv('data/data1.csv')
df.dropna(inplace=True)

In [20]:
df

,Culmen Length (mm),Culmen Depth (mm),Flipper Length (mm),Body Mass (g),Delta 15 N (o/oo),Delta 13 C (o/oo),Species
1,39.5,17.4,186.0,3800.0,8.94956,-24.69454,Adelie
2,40.3,18.0,195.0,3250.0,8.36821,-25.33302,Adelie
4,36.7,19.3,193.0,3450.0,8.76651,-25.32426,Adelie
5,39.3,20.6,190.0,3650.0,8.66496,-25.29805,Adelie
6,38.9,17.8,181.0,3625.0,9.18718,-25.21799,Adelie
...,...,...,...,...,...,...,...
338,47.2,13.7,214.0,4925.0,7.99184,-26.20538,Gentoo
340,46.8,14.3,215.0,4850.0,8.41151,-26.13832,Gentoo
341,50.4,15.7,222.0,5750.0,8.30166,-26.04117,Gentoo
342,45.2,14.8,212.0,5200.0,8.24246,-26.11969,Gentoo


# Q2

## What kind of data is in the column named "Species"?
 - quantitative ?
 - nominal ?
 - ordinal ?

### Ans: 

# Q3

## What kind of data is in the column named "Body Mass (g)"?
 - quantitative ?
 - nominal ?
 - ordinal ?

### Ans: 

# Q4

## The code below produces a new table. How many rows will the table have? (See the dataset description above for useful information).


### `df.groupby('Species')['Culmen Length (mm)'].aggregate([np.mean, np.median])`

### Ans: 

In [21]:
df.groupby('Species')['Culmen Length (mm)'].aggregate([np.mean, np.median])

/var/folders/zf/94x0k5092zgfxzfr4p5680gc0000gn/T/ipykernel_10825/2803495559.py:1: FutureWarning: The provided callable <function mean at 0x10f58ec20> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  df.groupby('Species')['Culmen Length (mm)'].aggregate([np.mean, np.median])
/var/folders/zf/94x0k5092zgfxzfr4p5680gc0000gn/T/ipykernel_10825/2803495559.py:1: FutureWarning: The provided callable <function median at 0x10f6d7640> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  df.groupby('Species')['Culmen Length (mm)'].aggregate([np.mean, np.median])


,mean,median
Species,,
Adelie,38.801418,38.80
Chinstrap,48.788060,49.50
Gentoo,47.506557,47.25


# Q5

## The following function accepts a series of numbers and returns a series of the same size in which every number has been standardized (also known as "z-scored"):

In [22]:
def standardize(ser):
    return (ser - ser.mean()) / ser.std()


### Which of the following lines of code will return a series of standardized flipper lengths of length 330 (same as the number of rows in df ), where the standardization is done within each species?

### That is, the flipper sizes of Adelie penguins are standardized as a group, the flipper sizes of Gentoo penguins are standardized as another group, and so on.

### 1.  `df.groupby('Species')['Flipper Length (mm)'].transform(standardize)`
<!-- &nbsp;&nbsp; -->
### 2.  `df.groupby('Species')['Flipper Length (mm)'].aggregate(standardize)`
<!-- &nbsp;&nbsp; -->
### 3.  `df.groupby('Species')['Flipper Length (mm)'].standardize()`
<!-- &nbsp;&nbsp; -->
### 4.  `df.groupby('Species')['Flipper Length (mm)'].agg(standardize)`

### Ans: 

In [23]:
df.groupby('Species')['Flipper Length (mm)'].transform(standardize)

1     -0.662246
2      0.726829
4      0.418146
5     -0.044880
6     -1.433955
         ...   
338   -0.484470
340   -0.330550
341    0.746891
342   -0.792310
343   -0.638390
Name: Flipper Length (mm), Length: 330, dtype: float64

In [24]:
# df.groupby('Species')['Flipper Length (mm)'].aggregate(standardize)

In [25]:
# df.groupby('Species')['Flipper Length (mm)'].standardize()

In [26]:
# df.groupby('Species')['Flipper Length (mm)'].agg(standardize)

# Q6

## Suppose incomplete data is collected on three new penguins, shown below:

In [27]:
df_incomplete = pd.DataFrame({
    'Flipper Length (mm)':[191.0,193.0,197.0],
    'Culmen Depth (mm)':[20.3,17.7,19.5],
    'Species': ['Adelie','Adelie','Adelie']
})
df_incomplete

,Flipper Length (mm),Culmen Depth (mm),Species
0,191.0,20.3,Adelie
1,193.0,17.7,Adelie
2,197.0,19.5,Adelie


### Call this dataframe `df_incomplete`, and note that there are fewer columns than in `df` and that the columns which do appear are in a different order.

### Suppose `df_incomplete` is appended to the end of `df` with the code `pd.concat([df, df_incomplete])`. What will be the values in the last line of the resulting dataframe?

#### 1.     [197, 19.5, 'Adelie', nan, nan, nan, nan] with row label (index) 2 
#### 2.     [nan, 19.5, 197.0, nan, nan, nan, 'Adelie'] with row label (index) 2 
#### 3.     [0, 19.5, 197.0, 0, 0, 0, 'Adelie'] with row label (index) 2 
#### 4.     [197, 19.5, 'Adelie', 0, 0, 0, 0] with row label (index) 2 
#### 5.     [197, 19.5, 'Adelie', nan, nan, nan, nan] with row label (index) 332 
#### 6.     [nan, 19.5, 197.0, nan, nan, nan, 'Adelie'] with row label (index) 332 
#### 7.     [0, 19.5, 197.0, 0, 0, 0, 'Adelie'] with row label (index) 332 
#### 8.     [197, 19.5, 'Adelie', 0, 0, 0, 0] with row label (index) 332

### Ans: 

In [28]:
df_concat = pd.concat([df,df_incomplete])
df_concat.tail()

,Culmen Length (mm),Culmen Depth (mm),Flipper Length (mm),Body Mass (g),Delta 15 N (o/oo),Delta 13 C (o/oo),Species
342,45.2,14.8,212.0,5200.0,8.24246,-26.11969,Gentoo
343,49.9,16.1,213.0,5400.0,8.36390,-26.15531,Gentoo
0,NaN,20.3,191.0,NaN,NaN,NaN,Adelie
1,NaN,17.7,193.0,NaN,NaN,NaN,Adelie
2,NaN,19.5,197.0,NaN,NaN,NaN,Adelie


# Q7

### Suppose a penguin is considered "little" if its body mass is less than 3250 grams. Write a piece of code that computes the number of each species which are considered little. Your code should return a series. To receive full credit, your code should not modify the dataframe df .

### Ans: 

In [29]:
df

,Culmen Length (mm),Culmen Depth (mm),Flipper Length (mm),Body Mass (g),Delta 15 N (o/oo),Delta 13 C (o/oo),Species
1,39.5,17.4,186.0,3800.0,8.94956,-24.69454,Adelie
2,40.3,18.0,195.0,3250.0,8.36821,-25.33302,Adelie
4,36.7,19.3,193.0,3450.0,8.76651,-25.32426,Adelie
5,39.3,20.6,190.0,3650.0,8.66496,-25.29805,Adelie
6,38.9,17.8,181.0,3625.0,9.18718,-25.21799,Adelie
...,...,...,...,...,...,...,...
338,47.2,13.7,214.0,4925.0,7.99184,-26.20538,Gentoo
340,46.8,14.3,215.0,4850.0,8.41151,-26.13832,Gentoo
341,50.4,15.7,222.0,5750.0,8.30166,-26.04117,Gentoo
342,45.2,14.8,212.0,5200.0,8.24246,-26.11969,Gentoo


In [44]:
df[df['Body Mass (g)'] < 3250].groupby('Species').size()

Species
Adelie       23
Chinstrap     3
dtype: int64

# Q8
### Suppose you have created a new table by_size which lists each penguin's size as either "small", "medium", or "large". The species is also included. The table looks like this:

<img src = "imgs/sizes.png"></img>

### Using the code `by_size.pivot_table(index='species', columns='size?', aggfunc='size')` , you've created the pivot table shown below:
<img src = 'imgs/pt.png'></img>

### There is a `NaN`in this pivot table. Why?

#### 1. some Gentoo penguins had sizes that were NaN
#### 2. some Gentoo penguins had sizes that were not strings
#### 3. there were too many small Gentoo penguins for the computer to represent with finite precision
#### 4. there were no small Gentoo penguins

### ans  4. there were no small Gentoo penguins.

Reasoning: NaN in a pivot table with aggfunc='size' indicates an empty intersection for those categories; no data points met both criteria.

# Q9 -

### Suppose we wish to fill in the missing value in the pivot table above. What makes the most sense to fill it in with?

#### A. 0
#### B. 1
#### C. np.inf (infinity)
#### D. the mean of the "small" column
#### E. the mean of the "Gentoo" row
#### F. a number chosen at random from the observed values in the "small" column
#### G. a number at random from the uniform distribution on 0, 1

### Ans: 

# Q10 - 1 Point

### You have a second dataframe, `personalities`, that contains the disposition and intelligence of several species of penguin:
<img src = 'imgs/personalities.png'></img>

In [45]:
personalities = pd.DataFrame({
    'Disposition':['happy','grumpy','mean'],
    'Intelligence':['smart','not smart','not smart'],
    'Species': ['Adelie','Chinstrap','Emperor']
})
personalities

,Disposition,Intelligence,Species
0,happy,smart,Adelie
1,grumpy,not smart,Chinstrap
2,mean,not smart,Emperor


### Note that personalities is missing some of the species that are in `df`, and it contains species that are not in `df`. You'd like to use the table to add columns to `df` for the disposition and intelligence of each of the penguins in the data set. 

### If a species in `df` is not listed in `personalities`, you're OK with having their disposition and Intelligence be `NaN`. 

### And if a species in `personalities` is not in `df`, it should not be added. 
### Therefore, the result of your merge should have exactly 330 rows -- the same as `df`.

### Which of the below will perform this?

#### 1. `pd.merge(df, personalities, how='outer')`
#### 2. `pd.merge(df, personalities, how='inner')`
#### 3. `pd.merge(df, personalities, how='left')`
#### 4. `pd.merge(df, personalities, how='right')`

### Ans: 

In [31]:
pd.merge(df, personalities, how='outer')

,Culmen Length (mm),Culmen Depth (mm),Flipper Length (mm),Body Mass (g),Delta 15 N (o/oo),Delta 13 C (o/oo),Species,Disposition,Intelligence
0,39.5,17.4,186.0,3800.0,8.94956,-24.69454,Adelie,happy,smart
1,40.3,18.0,195.0,3250.0,8.36821,-25.33302,Adelie,happy,smart
2,36.7,19.3,193.0,3450.0,8.76651,-25.32426,Adelie,happy,smart
3,39.3,20.6,190.0,3650.0,8.66496,-25.29805,Adelie,happy,smart
4,38.9,17.8,181.0,3625.0,9.18718,-25.21799,Adelie,happy,smart
...,...,...,...,...,...,...,...,...,...
326,47.2,13.7,214.0,4925.0,7.99184,-26.20538,Gentoo,NaN,NaN
327,46.8,14.3,215.0,4850.0,8.41151,-26.13832,Gentoo,NaN,NaN
328,50.4,15.7,222.0,5750.0,8.30166,-26.04117,Gentoo,NaN,NaN
329,45.2,14.8,212.0,5200.0,8.24246,-26.11969,Gentoo,NaN,NaN


In [32]:
pd.merge(df, personalities, how='inner')

,Culmen Length (mm),Culmen Depth (mm),Flipper Length (mm),Body Mass (g),Delta 15 N (o/oo),Delta 13 C (o/oo),Species,Disposition,Intelligence
0,39.5,17.4,186.0,3800.0,8.94956,-24.69454,Adelie,happy,smart
1,40.3,18.0,195.0,3250.0,8.36821,-25.33302,Adelie,happy,smart
2,36.7,19.3,193.0,3450.0,8.76651,-25.32426,Adelie,happy,smart
3,39.3,20.6,190.0,3650.0,8.66496,-25.29805,Adelie,happy,smart
4,38.9,17.8,181.0,3625.0,9.18718,-25.21799,Adelie,happy,smart
...,...,...,...,...,...,...,...,...,...
203,55.8,19.8,207.0,4000.0,9.70465,-24.53494,Chinstrap,grumpy,not smart
204,43.5,18.1,202.0,3400.0,9.37608,-24.40753,Chinstrap,grumpy,not smart
205,49.6,18.2,193.0,3775.0,9.46180,-24.70615,Chinstrap,grumpy,not smart
206,50.8,19.0,210.0,4100.0,9.98044,-24.68741,Chinstrap,grumpy,not smart


In [33]:
pd.merge(df, personalities, how='left')

,Culmen Length (mm),Culmen Depth (mm),Flipper Length (mm),Body Mass (g),Delta 15 N (o/oo),Delta 13 C (o/oo),Species,Disposition,Intelligence
0,39.5,17.4,186.0,3800.0,8.94956,-24.69454,Adelie,happy,smart
1,40.3,18.0,195.0,3250.0,8.36821,-25.33302,Adelie,happy,smart
2,36.7,19.3,193.0,3450.0,8.76651,-25.32426,Adelie,happy,smart
3,39.3,20.6,190.0,3650.0,8.66496,-25.29805,Adelie,happy,smart
4,38.9,17.8,181.0,3625.0,9.18718,-25.21799,Adelie,happy,smart
...,...,...,...,...,...,...,...,...,...
325,47.2,13.7,214.0,4925.0,7.99184,-26.20538,Gentoo,NaN,NaN
326,46.8,14.3,215.0,4850.0,8.41151,-26.13832,Gentoo,NaN,NaN
327,50.4,15.7,222.0,5750.0,8.30166,-26.04117,Gentoo,NaN,NaN
328,45.2,14.8,212.0,5200.0,8.24246,-26.11969,Gentoo,NaN,NaN


In [34]:
pd.merge(df, personalities, how='right')

,Culmen Length (mm),Culmen Depth (mm),Flipper Length (mm),Body Mass (g),Delta 15 N (o/oo),Delta 13 C (o/oo),Species,Disposition,Intelligence
0,39.5,17.4,186.0,3800.0,8.94956,-24.69454,Adelie,happy,smart
1,40.3,18.0,195.0,3250.0,8.36821,-25.33302,Adelie,happy,smart
2,36.7,19.3,193.0,3450.0,8.76651,-25.32426,Adelie,happy,smart
3,39.3,20.6,190.0,3650.0,8.66496,-25.29805,Adelie,happy,smart
4,38.9,17.8,181.0,3625.0,9.18718,-25.21799,Adelie,happy,smart
...,...,...,...,...,...,...,...,...,...
204,43.5,18.1,202.0,3400.0,9.37608,-24.40753,Chinstrap,grumpy,not smart
205,49.6,18.2,193.0,3775.0,9.46180,-24.70615,Chinstrap,grumpy,not smart
206,50.8,19.0,210.0,4100.0,9.98044,-24.68741,Chinstrap,grumpy,not smart
207,50.2,18.7,198.0,3775.0,9.39305,-24.25255,Chinstrap,grumpy,not smart


In [35]:
df # Reminder that merge operations are not automatically inPlace

,Culmen Length (mm),Culmen Depth (mm),Flipper Length (mm),Body Mass (g),Delta 15 N (o/oo),Delta 13 C (o/oo),Species
1,39.5,17.4,186.0,3800.0,8.94956,-24.69454,Adelie
2,40.3,18.0,195.0,3250.0,8.36821,-25.33302,Adelie
4,36.7,19.3,193.0,3450.0,8.76651,-25.32426,Adelie
5,39.3,20.6,190.0,3650.0,8.66496,-25.29805,Adelie
6,38.9,17.8,181.0,3625.0,9.18718,-25.21799,Adelie
...,...,...,...,...,...,...,...
338,47.2,13.7,214.0,4925.0,7.99184,-26.20538,Gentoo
340,46.8,14.3,215.0,4850.0,8.41151,-26.13832,Gentoo
341,50.4,15.7,222.0,5750.0,8.30166,-26.04117,Gentoo
342,45.2,14.8,212.0,5200.0,8.24246,-26.11969,Gentoo


### Ans: 

# Q14
### Suppose some of the entries in the "Body Mass (g)" column are missing.


### Body mass measurements were taken any time a penguin walked across a scale that was placed outside. In your experience, the scale sometimes fails to make a measurement if the item placed on top is too small.

### What is the most likely type of missingness in this case?

#### 1. Missing Completely at Random
#### 2. Missing at Random
#### 3. Not Missing at Random
#### 4. Missing by Design

### Ans: 

# Q15

### Suppose some of the entries in the "Delta 15 N (o/oo)" and "Delta 13 C (o/oo)" columns are missing.


### The measurements above were taken by four different teams located at different places in Antarctica, and one of the teams had a faulty measurement device that prevented them from measuring "Delta 15 N" and "Delta 13 N". The penguins in this team's area were mostly Gentoo penguins.

### What is the most likely type of missingness in this case?

#### 1. Missing Completely at Random
#### 2. Missing at Random
#### 3. Not Missing at Random
#### 4. Missing by Design

### Ans: 

# Q16

### Suppose some of the entries in the "Flipper Length (mm)" column are missing. It is very hard to measure the flipper length of a wild penguin because penguins have a tendency to run away whenever you approach them with a ruler. 

### For this reason, the researchers agree that they will only measure the flipper length of half of the penguins they see. To implement this, they flip a coin when they see a penguin -- if the coin comes up heads, they measure the penguin's flipper length, and if it is tails they leave the flipper length blank.

### What is the most likely type of missingness in this case?


#### 1. Missing Completely at Random
#### 2. Missing at Random
#### 3. Not Missing at Random
#### 4. Missing by Design

### Ans: 

# Q17

### Suppose some of the body masses are missing. You'd like to test if some of the body masses are missing at random (MAR) dependent on the species. 

### You will determine this with a permutation test. Which one of the below is a valid test statistic for this test?


#### 1. The TVD (total variation distance)
#### 2. The K-S statistic (Kolmogorov-Smirnov)
#### 3. The absolute (unsigned) difference in means
#### 4. The signed difference in means

To test if missingness depends on Species, we compare the categorical distribution of species when data is null vs. not null. TVD is the standard metric for comparing two categorical distributions.

# Q18


### Suppose the "Flipper Length (mm)" column has missing values. The distributions of species when the flipper length is missing (null) and when it is present (not null) are shown below:

<img src = 'imgs/missingness_distribution.png'></img>

### Futhermore, the average flipper length of each species is as follows:

In [36]:
df.groupby('Species')['Flipper Length (mm)'].mean()

Species
Adelie       190.290780
Chinstrap    195.671642
Gentoo       217.147541
Name: Flipper Length (mm), dtype: float64

### Suppose the overall mean flipper length is computed (using only the observed values).
### Which of the following is true?

### A. the mean will be biased high
### B. the mean will be biased low
### C. the mean will be unbiased

### Ans: The null distribution is heavily skewed toward Gentoo (~65%). Since Gentoo has the highest average flipper length (~217 mm compared to ~190-195 mm for others), missing their data disproportionately removes the highest values, dragging the observed mean down.


# Q19

### Write a piece of code that imputes missing flipper lengths with the median flipper length of the species that the penguin belongs to.

**Question 2**

Write a helper function, `midterm_review_imputation` that accepts `penguins_with_nulls` dataframe as an argument and returns a `pd.DataFrame` object where the cells in which `Flipper Length (mm)` is missing are **imputed with the median flipper length of the species** that the penguin belongs to. 



In [37]:
def midterm_review_imputation(penguins_with_nulls):
    """
    Takes a dataframe with NaN's or nulls and fills the `Flipper Length` column with the 
    Median Flipper length of the corresponding species.
    :Example:
    >>> penguins_with_nulls = pd.read_csv('data/data1.csv')
    >>> res = midterm_review_imputation(penguins_with_nulls)
    >>> isinstance(q2_out, pd.DataFrame)
    >>> q2_out['Flipper Length (mm)'].isnull().sum() == 0
    """
    df = penguins_with_nulls.copy(deep=True)
    # BEGIN SOLUTION
    df['Flipper Length (mm)']=df['Flipper Length (mm)'].fillna(df.groupby('Species')['Flipper Length (mm)'].transform('median'))
    # END SOLUTION
    return df